src/test/scala/is/hail/distributedmatrix/HailBlockMatrixSuite.scala

```diff
@@ -0,0 +1,522 @@
+package is.hail.distributedmatrix
+
+import breeze.linalg.{DenseMatrix => BDM}
+import is.hail.SparkSuite
+import is.hail.check.Arbitrary._
+import is.hail.check.Prop._
+import is.hail.check._
+import is.hail.distributedmatrix.HailBlockMatrix.ops._
+import is.hail.utils._
+import org.testng.annotations.Test
+
+class HailBlockMatrixSuite extends SparkSuite {
+
+  private val defaultBlockSize = 1024
+
+  def toBM(rows: Seq[Array[Double]]): HailBlockMatrix =
+    toBM(rows, defaultBlockSize)
+
+  def toBM(rows: Seq[Array[Double]], blockSize: Int): HailBlockMatrix = {
+    val n = rows.length
+    val m = if (n == 0) 0 else rows(0).length
+
+    HailBlockMatrix.from(sc, new BDM[Double](m, n, rows.flatten.toArray).t, blockSize)
+  }
+
+  def toBM(x: BDM[Double]): HailBlockMatrix =
+    toBM(x, defaultBlockSize)
+
+  def toBM(x: BDM[Double], blockSize: Int): HailBlockMatrix =
+    HailBlockMatrix.from(sc, x, blockSize)
+
+  def blockMatrixPreGen(blockSize: Int): Gen[HailBlockMatrix] = for {
+    (l, w) <- Gen.nonEmptySquareOfAreaAtMostSize
+    bm <- blockMatrixPreGen(l, w, blockSize)
+  } yield bm
+
+  def blockMatrixPreGen(rows: Int, columns: Int, blockSize: Int): Gen[HailBlockMatrix] = for {
+    arrays <- Gen.buildableOfN[Seq, Array[Double]](rows, Gen.buildableOfN(columns, arbDouble.arbitrary))
+  } yield toBM(arrays, blockSize)
+
+  val squareBlockMatrixGen = for {
+    size <- Gen.size
+    l <- Gen.interestingPosInt
+    s = math.sqrt(math.min(l, size)).toInt
+    blockSize <- Gen.interestingPosInt
+    bm <- blockMatrixPreGen(s, s, blockSize)
+  } yield bm
+
+  val blockMatrixGen = for {
+    blockSize <- Gen.interestingPosInt
+    bm <- blockMatrixPreGen(blockSize)
+  } yield bm
+
+  val twoMultipliableBlockMatrices = for {
+    Array(rows, inner, columns) <- Gen.nonEmptyNCubeOfVolumeAtMostSize(3)
+    blockSize <- Gen.interestingPosInt
+    x <- blockMatrixPreGen(rows, inner, blockSize)
+    y <- blockMatrixPreGen(inner, columns, blockSize)
+  } yield (x, y)
+
+  implicit val arbitraryHailBlockMatrix =
+    Arbitrary(blockMatrixGen)
+
+  @Test
+  def pointwiseSubtractCorrect() {
+    val m = toBM(Seq(
+      Array[Double](1,2,3,4),
```

Comment:
I'm inclined to not do this so that the tests are completely independent and also fully self-descriptive, but I will reformat everything to look like `multiplyByLocalMatrix`.